src/coreclr/src/pal/src/safecrt/wcslwr_s.cpp

```diff
@@ -0,0 +1,38 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+/***
+*wcslwr_s.cpp - contains _wcslwr_s() routine
+*
+*
+*Purpose:
+*   _wcslwr_s converts, in place, any upper case letters in input string to
+*   lowercase.
+*
+*******************************************************************************/
+
+#include <wctype.h>
+#include <string.h>
+#include <errno.h>
+#include <limits.h>
+#include "internal_securecrt.h"
+
+#include "mbusafecrt_internal.h"
+
+DLLEXPORT errno_t __cdecl _wcslwr_s(char16_t *string, size_t sz)
+{
+    _VALIDATE_RETURN_ERRCODE(string != NULL, EINVAL);
+    if (PAL_wcsnlen(string, sz) >= sz)
```

Comment:
@janvorli, I have tested this validation using the following program in VC++, to make sure it matches the expectation:

```c++
#include <wchar.h>
#include <stdio.h>

int main()
{
   wchar_t str[] = L"The String to End All Strings!";
   str[7] = NULL;

   // anything smaller than +1 is invalid: condition on line 13 is true
   // and _wcslwr_s on line 16 fails.
   size_t sz = wcslen(str) + 1;

   if (wcsnlen(str, sz) >= sz)
      printf("Error invalid length. wcsnlen(): %zd, sz: %zd\n", wcsnlen(str, sz), sz);
    
   errno_t err = _wcslwr_s(str, sz);

   if (err == 0)
   {
      wprintf(L"Success. result: %s\n", str);
   }
   else
   {
      printf("Error. number: %d\n", err);
   }
   return 0;
}
```

Actually, `_wcslwr_s` on line 16 with VC++ does not just fail, but it aborts the process during runtime without `+ 1` on line 11 (or with `- 1`, `- 2` and so on..). Seems like it does not gracefully validate the case: https://rextester.com/QXYW70578. In comparison, PAL implementation will fail gracefully in such case.

Reply:
I guess it depends of the failure callback. The _s functions can either return error or abort depending on it. It is described here: https://docs.microsoft.com/en-us/cpp/c-runtime-library/parameter-validation?view=vs-2019.

Reply:
Should we match Windows implementation and call abort()?

Reply:
It is actually there - the macros end up calling sMBUSafeCRTAssertFunc if it is set using MBUSafeCRTSetAssertFunc. We don't use it though for our runtime.

Reply:
I mean - that assert function can assert or abort or whater it wanted to do.